flang/lib/Lower/ConvertType.cpp

```diff
@@ -18,384 +18,184 @@
+static mlir::Type genComplexType(mlir::MLIRContext *context, int KIND) {
   if (Fortran::evaluate::IsValidKindOfIntrinsicType(
           Fortran::common::TypeCategory::Complex, KIND))
     return fir::ComplexType::get(context, KIND);
   return {};
 }
 
-namespace {
-
-/// Discover the type of an Fortran::evaluate::Expr<T> and convert it to an
-/// mlir::Type. The type returned may be an MLIR standard or FIR type.
-class TypeBuilder {
-public:
-  /// Constructor.
-  explicit TypeBuilder(
-      mlir::MLIRContext *context,
-      const Fortran::common::IntrinsicTypeDefaultKinds &defaults)
-      : context{context}, defaults{defaults} {}
-
-  //===--------------------------------------------------------------------===//
-  // Generate type entry points
-  //===--------------------------------------------------------------------===//
+static mlir::Type genFIRType(mlir::MLIRContext *context,
+                             Fortran::common::TypeCategory tc, int kind) {
+  switch (tc) {
+  case Fortran::common::TypeCategory::Real:
+    return genRealType(context, kind);
+  case Fortran::common::TypeCategory::Integer:
+    return genIntegerType(context, kind);
+  case Fortran::common::TypeCategory::Complex:
+    return genComplexType(context, kind);
+  case Fortran::common::TypeCategory::Logical:
+    return genLogicalType(context, kind);
+  case Fortran::common::TypeCategory::Character:
+    return genCharacterType(context, kind);
+  default:
+    break;
+  }
+  llvm_unreachable("unhandled type category");
+}
 
-  template <template <typename> typename A, Fortran::common::TypeCategory TC>
-  mlir::Type gen(const A<Fortran::evaluate::SomeKind<TC>> &) {
-    return genFIRType<TC>(context, defaultKind<TC>());
-  }
+//===--------------------------------------------------------------------===//
+// Symbol and expression type translation
+//===--------------------------------------------------------------------===//
 
-  template <template <typename> typename A, Fortran::common::TypeCategory TC,
-            int KIND>
-  mlir::Type gen(const A<Fortran::evaluate::Type<TC, KIND>> &) {
-    return genFIRType<TC, KIND>(context);
-  }
+/// TypeBuilder translates expression and symbol type taking into account
+/// their shape and length parameters. For symbols, attributes such as
+/// ALLOCATABLE or POINTER are reflected in the fir type.
+/// It uses evaluate::DynamicType and evaluate::Shape when possible to
+/// avoid re-implementing type/shape analysis here.
+namespace {
+struct TypeBuilder {
 
-  // breaks the conflict between A<Type<TC,KIND>> and Expr<B> deduction
-  template <Fortran::common::TypeCategory TC, int KIND>
-  mlir::Type
-  gen(const Fortran::evaluate::Expr<Fortran::evaluate::Type<TC, KIND>> &) {
-    return genFIRType<TC, KIND>(context);
-  }
+  mlir::Type genExprType(const Fortran::lower::SomeExpr &expr) {
+    auto dynamicType = expr.GetType();
+    if (!dynamicType)
+      llvm::report_fatal_error("lowering typeless expression type");
```

Comment:
How is the client supposed to deal with these? Should this be an `llvm::Optional` type return?

Reply:
Good point. The previous code would return `genTypelessPtr` (pointer to empty tuple) for all typless expression, but this was not used.
Typeless expressions are:
- BOZLiteralConstant -> let's use `NoneType` (I think these do not hit lowering since they are folded to whatever type they should be according to the context they appear in.)
- NullPointer -> let's use  `fir.ref<NoneType>` since that's what is used by [createNullConstant](https://github.com/flang-compiler/f18-llvm-project/blob/f75fcb9a706ba7b6cc22948ffbaf555c696c599b/flang/lib/Lower/FIRBuilder.cpp#L49).
- ProcedureDesignator -> That one is not typeless in FIR since we have function types. So we should return it.
- ProcedureRef -> subroutine calls return value, I guess `NoneType` could also be used here.

Again, right now there are no clients for those, so maybe a hard TODO is better until there is a use case and we can pick the type according to the use case need.

Reply:
I added handling for typeless expr in the latest commit. That required threading the `AbstractConverter`  so that we can use the call interface signature lowering on the `ProcedureDesignator`. But I took the opportunity of having this at hand to remove the foldingContext/mlirContext threading and lower symbol location for better mlir::emitError messages.